File: Algorithms/Reverse_Bits.py
'''



We are asked to reverse bits in our number. What is the most logical way to do it? 
Create number out, process original number bit by bit from end and 
add this bit to the end of our out number, and that is all! Why it is works?

out = (out << 1)^(n & 1) adds last bit of n to out
n >>= 1 removes last bit from n.
Imagine number n = 11011010, and out = 0

out = 0, n = 1101101
out = 01, n = 110110
out = 010, n = 11011
out = 0101, n = 1101
out = 01011, n = 110
out = 010110, n = 11
out = 0101101, n = 1
out = 01011011, n = 0
Compexity: time complexity is O(32), space complexity is O(1).




'''


class Solution:
    def reverseBits(self, n):
        out = 0
        for i in range(32):
            out = (out << 1)^(n & 1)
            n >>= 1
        return out
      
      

      
      

# Bit manipulation

def reverseBits(self, n):
    ans = 0
    for i in xrange(32):
        ans = (ans << 1) + (n & 1)
        n >>= 1
    return ans
  
  
  
  
  
  
'''

One pass for just 16 times.
Swapping the bit and or them into a new number.

'''

class Solution:
    def reverseBits(self, n):
        ret = 0
        for i in range(16):
            right_bit = (n>>i)&1
            left_bit = (n>>(31-i))&1
            ret |= right_bit<<(31-i)
            ret |= left_bit<<i
        return ret
      
      

      
      
      
      
'''


Divide and Conquer!  Someway like merge sort.
For example, if there are 8 bit binary number abcdefgh,
the process is as follow:
abcdefgh -> efghabcd -> ghefcdab -> hgfedcba


'''

class Solution:
    def reverseBits(self, n):
        # For python, there is no 32bit int, so we need to force it 32 bits.
        n = (n >> 16) | (n << 16) & 0xffffffff
        n = ((n & 0xff00ff00) >> 8) | ((n & 0x00ff00ff) << 8) & 0xffffffff
        n = ((n & 0xf0f0f0f0) >> 4) | ((n & 0x0f0f0f0f) << 4) & 0xffffffff
        n = ((n & 0xcccccccc) >> 2) | ((n & 0x33333333) << 2) & 0xffffffff
        n = ((n & 0xaaaaaaaa) >> 1) | ((n & 0x55555555) << 1) & 0xffffffff
        return n

      
      
  
  
  
      
# Compress the four lines into one

class Solution:
    def reverseBits(self, n):
        ret = 0
        for i in range(16):
            ret |= ((n>>i)&1)<<(31-i) | ((n>>(31-i))&1)<<i
        return ret
      
      
      
      
# minimised solution


class Solution:
    # @param n, an integer
    # @return an integer
    def reverseBits(self, n):
        oribin='{0:032b}'.format(n)
        reversebin=oribin[::-1]
        return int(reversebin,2)
      
      
      
# string manipulation, slicing


class Solution:
    def reverseBits(self, n):
        return int(bin(n)[2:].zfill(32)[::-1], 2)
      
      

      
      

File: Algorithms/Reverse_Bits.java
// Bitwise

public class Solution {
    // you need treat n as an unsigned value
    public int reverseBits(int n) {
        int result = 0;
        for (int i = 0; i < 32; i++) {
            result += n & 1;
            n >>>= 1;   // CATCH: must do unsigned shift
            if (i < 31) // CATCH: for last digit, don't shift!
                result <<= 1;
        }
        return result;
    }
}



// How to optimize if this function is called multiple times? We can divide an int into 4 bytes, 
// and reverse each byte then combine into an int. For each byte, we can use cache to improve performance.

public class Solution {
    // you need treat n as an unsigned value
    // cache
    private final Map<Byte, Integer> cache = new HashMap<Byte, Integer>();
    public int reverseBits(int n) {
        byte[] bytes = new byte[4];
        for (int i = 0; i < 4; i++) // convert int into 4 bytes
            bytes[i] = (byte)((n >>> 8*i) & 0xFF);
        int result = 0;
        for (int i = 0; i < 4; i++) {
            result += reverseByte(bytes[i]); // reverse per byte
            if (i < 3)
                result <<= 8;
        }
        return result;
    }

    private int reverseByte(byte b) {
        Integer value = cache.get(b); // first look up from cache
        if (value != null)
            return value;
        value = 0;
        // reverse by bit
        for (int i = 0; i < 8; i++) {
            value += ((b >>> i) & 1);
            if (i < 7)
                value <<= 1;
        }
        cache.put(b, value);
        return value;
    }
}



/*
More optimized

It seems that HashMap can accelerate the algorithm, but the fact is not as we hope.
If we repeat the function 10,000,000 times, the first algorithm you provide will cost 8,940 ms, 
the second algorithm you provide will cost 12,454 ms, while the following code will cost only 3,500 ms.

The optimized code runs slower.

I think reasons are as follows:
1.In the following code, the bit operation is QUITE FAST, 
  no if/while/for statment, thus no branch predictions, 
  and the efficiency of every statement will be almost 100%.

2.In the "cache" solution, the function calls、 
  brach predictions 、read/write-memory operations or 
  even hash collisions will all contribute to the cost of time.

*/



public class Solution {
    // you need treat n as an unsigned value
    // cache
    public int reverseBits(int n) {
        int ret=n;
        ret = ret >>> 16 | ret<<16;
        ret = (ret & 0xff00ff00) >>> 8 | (ret & 0x00ff00ff) << 8;
        ret = (ret & 0xf0f0f0f0) >>> 4 | (ret & 0x0f0f0f0f) << 4;
        ret = (ret & 0xcccccccc) >>> 2 | (ret & 0x33333333) << 2;
        ret = (ret & 0xaaaaaaaa) >>> 1 | (ret & 0x55555555) << 1;
        return ret;
    }
}




// Another solution

public class Solution {
    // you need treat n as an unsigned value
    public int reverseBits(int n) {
        if (n == 0) return 0;

        int result = 0;
        for (int i = 0; i < 32; i++) {
            result <<= 1;
            if ((n & 1) == 1) result++;
            n >>= 1;
        }
        return result;
    }
}







// Divide and Conquer


public class Solution {
    
    public int reverseBits(int num) {
        
        num = ((num & 0xffff0000) >>> 16) | ((num & 0x0000ffff) << 16);
        num = ((num & 0xff00ff00) >>> 8) | ((num & 0x00ff00ff) << 8);
        num = ((num & 0xf0f0f0f0) >>> 4) | ((num & 0x0f0f0f0f) << 4);
        num = ((num & 0xcccccccc) >>> 2) | ((num & 0x33333333) << 2);
        num = ((num & 0xaaaaaaaa) >>> 1) | ((num & 0x55555555) << 1);
        
        return num;
        
    }
}





// minimised solution



public class Solution {
    // you need treat n as an unsigned value
    public int reverseBits(int n) {
      int result = 0;
      for (int i = 0; i < 32; ++i) {
        result = result<<1  | (n & 1);
        n >>>= 1;
      }
    return result;
    }
}


